ssx: exit early from sleep_abortable if already aborted

nvartolomei opened the pull request:
Caught a few shutdown hangs in unit tests which I tracked down to sleep_abortable being invoked with high durations after abort sources where already aborted.

Do not try to sleep if abort source is already set. ss::sleep_abortable behaves similarly from what I can read.

```
sleeper(typename Clock::duration dur, abort_source& as)
...
auto sc_opt = as.subscribe // returns empty optional if abort source is
already set
...
if (sc_opt) {...} else {
    done.set_exception(sleep_aborted());
}
```

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v24.2.x
- [x] v24.1.x
- [ ] v23.3.x

## Release Notes
* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

nvartolomei commented:
This fixes hang in metadata uploader unit test (cc @andrwng). There seems to be a race where fixture is being destroyed and config is reset to default upload interval of 1h. Then abort source is set and uploader goes to sleep for 1h https://github.com/redpanda-data/redpanda/blob/4352d77d0d6169aec4db99bbce793e8cbb3cd36f/src/v/cluster/cloud_metadata/uploader.cc#L400

That sleep won't happen anymore.

andrwng reviewed (approved):
Nice find!

dotnwat commented on src/v/ssx/sleep_abortable.h:
```diff
@@ -44,6 +44,11 @@ sleep_abortable(typename Clock::duration dur, AbortSource&... src) {
     };
+
+    if ((src.abort_requested() || ...)) {
+        return seastar::make_exception_future<>(seastar::sleep_aborted());
```
ahh nice. maybe the assumption by the author was that subscribing to an aborted source would inline invoke the callback, in which case i think this would have worked correctly since presumably sleep_abortable would check the abort source before sleeping.

vbotbuildovich commented:
<!--RETRIED FAILURES-->
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/55330#0192348d-4617-400a-a713-3c235c94935e
<!--ba20d162-8d69-480f-ae5c-776334057181-->

vbotbuildovich commented:
<!--NEW FAILURES-->
new failures in https://buildkite.com/redpanda/redpanda/builds/55376#019237e5-3a9e-4137-ace0-4940126ba38f:
```
"rptest.tests.e2e_shadow_indexing_test.ShadowIndexingManyPartitionsTest.test_many_partitions_recovery"
```
<!--474cceee-7be9-4c4d-9077-fd5fc0afcc3e-->

piyushredpanda commented:
/ci-repeat 1

vbotbuildovich commented:
/backport v24.2.x

vbotbuildovich commented:
/backport v24.1.x

dotnwat commented:
> @[nvartolomei](https://github.com/nvartolomei) nvartolomei [force-pushed](https://github.com/redpanda-data/redpanda/compare/f487ce67b1ef141d1024a5d3b58fe3e341450c00..99379758291a15ee46f6df572ef0b0cfb1a6d9aa) the nv/sleep-abortable-hang branch from [f487ce6 ](https://github.com/redpanda-data/redpanda/commit/f487ce67b1ef141d1024a5d3b58fe3e341450c00)to 9937975

@nvartolomei why was there an extra force push here before merging?

nvartolomei commented:
@dotnwat most likely because full disk gc test failed and I thought it is because this branch didn't contain the fix. Rebased to get the commit in and get a green CI run. Looks like the commit was already here though and it failed for another reason which I fixed later in https://github.com/redpanda-data/redpanda/pull/23550

dotnwat commented:
> Rebased to get the commit in and get a green CI run

CI automatically rebases. The only time that is useful is if you are fixing a merge conflict.
